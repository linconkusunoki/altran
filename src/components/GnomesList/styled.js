import styled from 'styled-components'
import { FixedSizeList } from 'react-window'
import { Link as RouterLink } from 'react-router-dom'

export const Row = styled.div`
  :hover {
    background-color: rgba(14, 30, 37, 0.05);
  }
`

export const Card = styled.div`
  display: flex;
  align-items: center;
  overflow: hidden;
`

export const Details = styled.div`
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;

  @media (min-width: 768px) {
    flex: 1;
    display: flex;
    justify-content: space-between;
  }
`

export const Name = styled.h3`
  color: rgba(14, 30, 37, 0.87);
  font-size: 18px;
  line-height: 24px;
  margin: 0;
  text-overflow: ellipsis;
  overflow: hidden;

  @media (min-width: 768px) {
    flex-basis: 50%;
  }
`

export const InfoWrapper = styled.div`
  @media (min-width: 768px) {
    flex: 1;
    display: flex;
    flex-wrap: wrap;
  }
`

export const Info = styled.div`
  color: rgba(14, 30, 37, 0.54);
  white-space: nowrap;

  &:nth-child(odd) {
    display: none;
  }

  @media (min-width: 768px) {
    width: 50%;

    &:nth-child(odd) {
      display: block;
    }
  }
`

export const FiltersWrapper = styled.div`
  margin-bottom: 24px;

  > div {
    display: flex;
    flex: 1;
    align-items: flex-end;
  }

  img {
    height: 50px;
  }

  @media (min-width: 768px) {
    display: flex;
    justify-content: space-around;
    align-items: flex-start;

    > div {
      margin-left: 16px;

      &:first-child {
        margin-left: 0;
      }
    }

    img {
      height: 80px;
    }
  }
`

export const Search = styled.input`
  display: block;
  padding: 15px;
  border: 1px solid #ddd;
  border-radius: 6px;
  width: 100%;
  background: #f1f1f1;
  flex: 1;
`

export const SelectsWrapper = styled.div`
  flex: 1;

  @media (min-width: 768px) {
    display: flex;
    flex: 0.6;
  }
`

export const List = styled(FixedSizeList)`
  margin: 0 -24px;
`

export const Link = styled(RouterLink)`
  padding: 12px 24px;
  display: block;
  text-decoration: none;
`
