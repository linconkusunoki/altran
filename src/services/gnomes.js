import api from './api'

const get = async id => {
  let res

  try {
    res = await api.get('/rrafols/mobile_test/master/data.json')
  } catch (error) {
    return error
  }

  if (id) {
    const gnome = res.data.Brastlewark.find(gnome => gnome.id === parseInt(id))
    return gnome
  }

  return res.data.Brastlewark
}

export default {
  get,
}
