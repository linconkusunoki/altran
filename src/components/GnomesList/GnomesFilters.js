import React from 'react'

import Select from '../Select'
import * as S from './styled'

const GnomesFilters = ({ items, onFilter, onSearch, filters }) => {
  const [professions, setProfessions] = React.useState([])
  const [hairColors, setHairColors] = React.useState([])

  React.useEffect(() => {
    let newProfessions = []
    let newHairColors = []

    items.map(item => {
      // group all professions
      item.professions.map(profession => {
        if (!newProfessions.includes(profession)) {
          newProfessions.push(profession)
        }
        return null
      })

      // hair color
      if (!newHairColors.includes(item.hair_color)) {
        newHairColors.push(item.hair_color)
      }

      return null
    })

    setProfessions(newProfessions)
    setHairColors(newHairColors)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  const handleChange = e => {
    const name = e.target.name
    const value = e.target.value
    onFilter({ [name]: value === '' ? null : value })
  }

  const handleSearch = e => {
    onSearch(e.target.value)
  }

  return (
    <S.FiltersWrapper>
      <S.Search
        type="search"
        placeholder="Search name"
        onChange={handleSearch}
      />
      <S.SelectsWrapper>
        <Select
          title="Professions"
          name="professions"
          options={professions}
          onChange={handleChange}
          value={filters.professions}
        />
        <Select
          title="Hair colors"
          name="hair_color"
          options={hairColors}
          onChange={handleChange}
          value={filters.hair_color}
        />
      </S.SelectsWrapper>
    </S.FiltersWrapper>
  )
}

export default GnomesFilters
