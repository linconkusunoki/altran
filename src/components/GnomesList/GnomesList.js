import React from 'react'

import gnomesService from '../../services/gnomes'
import useFetch from '../../hooks/useFetch'
import GnomesFilters from './GnomesFilters'
import Avatar from '../Avatar'
import HairColor from '../HairColor'
import * as S from './styled'

const GnomesList = () => {
  const initialFilters = { professions: null, hair_color: null }
  const { isLoading, error, response } = useFetch(gnomesService.get)
  const [items, setItems] = React.useState(false)
  const [filteredHairColors, setFilteredHairColors] = React.useState([])
  const [filteredProfessions, setFilteredProfessions] = React.useState([])
  const [filters, setFilter] = React.useState(initialFilters)

  React.useEffect(() => {
    setItems(response)
  }, [response])

  React.useEffect(() => {
    if (items) {
      Object.keys(filters).map(filter => {
        const applied = filters[filter]
        if (!applied) return null

        // profession
        if (filter === 'professions') {
          const professionsFilter = items.filter(item => {
            return item.professions.includes(applied)
          })
          setFilteredProfessions([...professionsFilter])
        }

        // hair color
        if (filter === 'hair_color') {
          const hairFilter = response.filter(item => {
            return item.hair_color === applied
          })
          setFilteredHairColors([...hairFilter])
        }

        return null
      })
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [filters])

  const getItemsToRender = () => {
    const source = [...new Set([...filteredHairColors, ...filteredProfessions])]
    let filtered

    if (!filters.professions && !filters.hair_color) {
      return { list: items, size: items.length }
    }

    if (filters.professions) {
      filtered = source.filter(item => {
        return item.professions.includes(filters.professions)
      })
    }

    if (filters.hair_color) {
      if (filtered) {
        filtered = filtered.filter(item => {
          return item.hair_color === filters.hair_color
        })
      } else {
        filtered = response.filter(item => {
          return item.hair_color === filters.hair_color
        })
      }
    }

    return { list: filtered, size: filtered.length }
  }

  const handleFilter = newFilter => {
    setFilter({ ...filters, ...newFilter })
  }

  const handleSearch = value => {
    setFilter(initialFilters)
    if (value.length === 0) {
      return setItems(response)
    }

    if (value.length >= 3) {
      const newItems = response.filter(item => {
        return item.name.toLowerCase().indexOf(value) > -1
      })
      setItems(newItems)
    }
  }

  if (isLoading) {
    return <p>Loading gnomes...</p>
  }

  if (error) {
    return <p>Error while loading gnomes!</p>
  }

  const { list, size } = getItemsToRender()

  const Row = ({ index, style }) => {
    const gnome = list[index]
    return (
      <S.Row style={style}>
        <S.Link to={`/gnome/${gnome.id}`}>
          <S.Card>
            <Avatar thumbnail={gnome.thumbnail} />
            <S.Details>
              <S.Name>{gnome.name}</S.Name>
              <S.InfoWrapper>
                <S.Info>height: {gnome.height.toFixed(2)}</S.Info>
                <S.Info>
                  hair color: <HairColor color={gnome.hair_color} />
                </S.Info>
                <S.Info>weight: {gnome.weight.toFixed(2)}</S.Info>
                <S.Info>age: {gnome.age}</S.Info>
              </S.InfoWrapper>
            </S.Details>
          </S.Card>
        </S.Link>
      </S.Row>
    )
  }

  return (
    <div>
      <GnomesFilters
        items={list}
        onFilter={handleFilter}
        onSearch={handleSearch}
        filters={filters}
      />
      {size === 0 ? 'No gnome found :(' : `Gnomes found: ${size}`}
      <S.List height={500} itemCount={size} itemSize={94}>
        {Row}
      </S.List>
    </div>
  )
}

export default GnomesList
