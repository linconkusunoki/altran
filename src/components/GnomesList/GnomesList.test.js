import React from 'react'
import { MemoryRouter } from 'react-router-dom'
import { render, wait } from '@testing-library/react'
import GnomesList from './'
import gnomes from '../../services/gnomes'

jest.mock('../../services/gnomes')

const gnome = {
  age: 306,
  friends: ['Cogwitz Chillwidget', 'Tinadette Chillbuster'],
  hair_color: 'Pink',
  height: 107.75835,
  id: 0,
  name: 'Tobus Quickwhistle',
  professions: [
    'Metalworker',
    'Woodcarver',
    'Stonecarver',
    ' Tinker',
    'Tailor',
    'Potter',
  ],
  thumbnail:
    'http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg',
  weight: 39.065952,
}

it('should render correctly', async () => {
  gnomes.get.mockResolvedValueOnce([gnome])
  const { container, getByText, getByTestId } = render(
    <MemoryRouter>
      <GnomesList />
    </MemoryRouter>,
  )

  expect(getByText('Loading gnomes...')).toBeTruthy()
  await wait(() => expect(getByTestId('amount')).toBeTruthy())
  expect(container.firstChild).toMatchSnapshot()
})

it('should render an empty list', async () => {
  gnomes.get.mockResolvedValueOnce([])
  const { getByTestId, getByText } = render(
    <MemoryRouter>
      <GnomesList />
    </MemoryRouter>,
  )
  expect(getByText('Loading gnomes...')).toBeTruthy()
  await wait(() => expect(getByTestId('amount')).toBeTruthy())
  expect(getByTestId('amount').textContent).toEqual('No gnome found :(')
})

it('should render an error message', async () => {
  gnomes.get.mockRejectedValueOnce(new Error('Async error'))
  const { getByText } = render(
    <MemoryRouter>
      <GnomesList />
    </MemoryRouter>,
  )
  expect(getByText('Loading gnomes...')).toBeTruthy()
  await wait(() => expect(getByText('Error while loading gnomes!')))
})
