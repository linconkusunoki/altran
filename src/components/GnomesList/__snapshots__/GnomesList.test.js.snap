// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div
    class="sc-dnqmqq hUhyYx"
  >
    <div>
      <img
        alt="gnome"
        height="50"
        src="gnome.svg"
      />
      <input
        class="sc-iwsKbI iqGaBW"
        data-testid="search"
        placeholder="Search name"
        type="search"
        value=""
      />
    </div>
    <div>
      <img
        alt="hammer"
        height="50"
        src="hammer.svg"
      />
      <div
        class="sc-bdVaJa iUbJjP"
      >
        <select
          class="sc-bwzfXH fiTiQi"
          data-testid="select-professions"
          name="professions"
        >
          <option
            value=""
          >
            Professions
          </option>
          <option
            value="Metalworker"
          >
            Metalworker
          </option>
          <option
            value="Woodcarver"
          >
            Woodcarver
          </option>
          <option
            value="Stonecarver"
          >
            Stonecarver
          </option>
          <option
            value=" Tinker"
          >
             Tinker
          </option>
          <option
            value="Tailor"
          >
            Tailor
          </option>
          <option
            value="Potter"
          >
            Potter
          </option>
        </select>
      </div>
    </div>
    <div>
      <img
        alt="potion"
        height="50"
        src="potion.svg"
      />
      <div
        class="sc-bdVaJa iUbJjP"
      >
        <select
          class="sc-bwzfXH fiTiQi"
          data-testid="select-hair"
          name="hair_color"
        >
          <option
            value=""
          >
            Hair colors
          </option>
          <option
            value="Pink"
          >
            Pink
          </option>
        </select>
      </div>
    </div>
  </div>
  <p
    data-testid="amount"
  >
    Gnomes found: 1
  </p>
  <div
    class="sc-gqjmRU bbbUrX"
    style="position: relative; height: 500px; overflow: auto; will-change: transform; direction: ltr;"
  >
    <div
      style="height: 94px; width: 100%;"
    >
      <div
        class="sc-bxivhb biguMO"
        data-testid="item"
        style="position: absolute; left: 0px; top: 0px; height: 94px; width: 100%;"
      >
        <a
          class="sc-VigVT bShVlT"
          href="/gnome/0"
        >
          <div
            class="sc-ifAKCX cXCwHH"
          >
            <figure
              class="sc-jTzLTM jsahtH"
            />
            <div
              class="sc-EHOje jqPQKJ"
            >
              <h3
                class="sc-bZQynM jkNGcG"
              >
                Tobus Quickwhistle
              </h3>
              <div
                class="sc-gzVnrw fcOYkR"
              >
                <div
                  class="sc-htoDjs mANFE"
                >
                  height: 
                  107.76
                </div>
                <div
                  class="sc-htoDjs mANFE"
                >
                  hair color: 
                  <span
                    class="sc-jzJRlG ekAJnd"
                    color="Pink"
                  />
                </div>
                <div
                  class="sc-htoDjs mANFE"
                >
                  weight: 
                  39.07
                </div>
                <div
                  class="sc-htoDjs mANFE"
                >
                  age: 
                  306
                </div>
              </div>
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</div>
`;
